ports/esp32/machine_encoder.c

```diff
@@ -0,0 +1,787 @@
+See also
+https://github.com/espressif/esp-idf/tree/master/examples/peripherals/pcnt/rotary_encoder
+*/
+
+#include "py/runtime.h"
+#include "mphalport.h"
+#include "modmachine.h"
+
+#if MICROPY_PY_MACHINE_PCNT
+
+#include "driver/pcnt.h"
+#include "soc/pcnt_struct.h"
+#include "esp_err.h"
+
+#include "machine_encoder.h"
+
+#define GET_INT mp_obj_get_int_truncated
+// #define GET_INT mp_obj_get_ll_int // need PR: py\obj.c: Get 64-bit integer arg. #80896
+
+STATIC pcnt_isr_handle_t pcnt_isr_handle = NULL;
+STATIC mp_pcnt_obj_t *pcnts[PCNT_UNIT_MAX] = {};
+
+#if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4, 1, 0)
+#define EVT_THRES_0 PCNT_EVT_THRES_0
+#define EVT_THRES_1 PCNT_EVT_THRES_1
+#define EVT_ZERO    PCNT_EVT_ZERO
+#else
+#define EVT_THRES_0 (1 << PCNT_EVT_THRES_0)
+#define EVT_THRES_1 (1 << PCNT_EVT_THRES_1)
+#define EVT_ZERO    (1 << PCNT_EVT_ZERO)
+#endif
+
+/* Decode what PCNT's unit originated an interrupt
+ * and pass this information together with the event type
+ * the main program using a queue.
+ */
+#if CONFIG_IDF_TARGET_ESP32S2 || CONFIG_IDF_TARGET_ESP32S3
+#define H_LIM_LAT  cnt_thr_h_lim_lat_un
+#define L_LIM_LAT  cnt_thr_l_lim_lat_un
+#define THRES0_LAT cnt_thr_thres0_lat_un
+#define THRES1_LAT cnt_thr_thres0_lat_un
+#define ZERO_LAT   cnt_thr_zero_lat_un
+#else
+#define H_LIM_LAT  h_lim_lat
+#define L_LIM_LAT  l_lim_lat
+#define THRES0_LAT thres0_lat
+#define THRES1_LAT thres1_lat
+#define ZERO_LAT   zero_lat
+#endif
+STATIC void IRAM_ATTR pcnt_intr_handler(void *arg) {
+    for (int id = 0; id < PCNT_UNIT_MAX; ++id) {
+        if (PCNT.int_st.val & (1 << id)) {
+            mp_pcnt_obj_t *self = pcnts[id];
+            if (self != NULL) {
+                if (PCNT.status_unit[id].H_LIM_LAT) {
+                    self->counter += INT16_ROLL;
+                } else if (PCNT.status_unit[id].L_LIM_LAT) {
+                    self->counter -= INT16_ROLL;
+                }
+
+                self->status = 0;
+                if (PCNT.status_unit[id].THRES1_LAT) {
+                    if (self->counter == self->counter_match1) {
+                        self->status |= EVT_THRES_1;
+                        mp_sched_schedule(self->handler_match1, MP_OBJ_FROM_PTR(self));
+                        mp_hal_wake_main_task_from_isr();
+                    }
+                }
+                if (PCNT.status_unit[id].THRES0_LAT) {
+                    if (self->counter == self->counter_match2) {
+                        self->status |= EVT_THRES_0;
+                        mp_sched_schedule(self->handler_match2, MP_OBJ_FROM_PTR(self));
+                        mp_hal_wake_main_task_from_isr();
+                    }
+                }
+                if (PCNT.status_unit[id].ZERO_LAT) {
+                    if (self->counter == 0) {
+                        self->status |= EVT_ZERO;
+                        mp_sched_schedule(self->handler_zero, MP_OBJ_FROM_PTR(self));
```

Comment:
How is this supposed to be used by the user? This triggers when the hardware counter passes zero. But the user setting the counter value or limit overflows will set self->counter to values != 0. So from the users perspective this will fire at counter values != 0 as the value the user sees includes self->counter.